keras_contrib/optimizers/yogi.py

```diff
@@ -0,0 +1,92 @@
+from keras import backend as K
+from keras.optimizers import Optimizer
+from keras.utils.generic_utils import get_custom_objects
+
+
+class Yogi(Optimizer):
+    """Yogi optimizer.
+    Yogi is a variation of Adam that controls the increase in effective
+    learning rate, which (according to the paper) leads to even better
+    performance than Adam with similar theoretical guarantees on convergence.
+    Default parameters follow those provided in the original paper, Tab.1
+    # Arguments
+        lr: float >= 0. Learning rate.
+        beta_1: float, 0 < beta < 1. Generally close to 1.
+        beta_2: float, 0 < beta < 1. Generally close to 1.
+        epsilon: float >= 0. Fuzz factor. If `None`, defaults to `K.epsilon()`.
+        decay: float >= 0. Learning rate decay over each update.
+    # References
+        - [Adaptive Methods for Nonconvex Optimization](
+           https://papers.nips.cc/paper/8186-adaptive-methods-for-nonconvex-optimization)
+
+    If you open a issues or a pull request about the Yogi optimizer,
+    please add 'cc @MarcoAndreaBuchmann' to notify him.
+    """
+
+    def __init__(self, lr=0.01, beta_1=0.9, beta_2=0.999,
+                 epsilon=1e-3, decay=0., **kwargs):
```

Comment:
Is there a strong reason why epsilon can default to `K.epsilon()`? In keras, all optimizers have the default epsilon to `K.epsilon()`.

Reply:
Thanks for having such a close look at the implementation.

In the [paper](https://papers.nips.cc/paper/8186-adaptive-methods-for-nonconvex-optimization), `epsilon` is `1e-2` (Tab 1) or `1e-3` (Tab 2), whereas the backend's `epsilon` is orders of magnitude smaller (e.g.`1e-7` for [tensorflow](https://www.tensorflow.org/api_docs/python/tf/keras/backend/epsilon)). In my opinion, providing this default value is therefore necessary, as not doing so would mean that users would have a parameter that's off by orders of magnitude from the values in the paper it's meant to reproduce.

I understand your point of preferring consistency across optimizers, but in this case it would not lead to a faithful reproduction of the paper.

Reply:
Ok it makes sense. Let's keep it like that.